<a href="https://colab.research.google.com/github/amii-uwl-dba/northstar-analytics/blob/main/03_mongodb.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [2]:
# Mount Drive
from google.colab import drive
drive.mount('/content/drive')

print("Drive mounted successfully")

Mounted at /content/drive
Drive mounted successfully


In [3]:
import zipfile
import os
import pandas as pd

zip_path = '/content/drive/MyDrive/northstar_dataset.zip'
extract_path = '/content/northstar_dataset/'

with zipfile.ZipFile(zip_path, 'r') as z:
    z.extractall(extract_path)

path = '/content/northstar_dataset/northstar_dataset/'

customers  = pd.read_csv(path + 'customers.csv')
orders     = pd.read_csv(path + 'orders.csv')
deliveries = pd.read_csv(path + 'deliveries.csv')
complaints = pd.read_csv(path + 'complaints.csv')
app_events = pd.read_csv(path + 'app_events.csv')

print("All files loaded successfully")

All files loaded successfully


In [4]:
!pip install pymongo dnspython

from pymongo import MongoClient

client = MongoClient("mongodb+srv://amiimiyuu2098_db_user:GOBFo62eG9JpPEte@northstar-cluster.qqanxwn.mongodb.net/?appName=northstar-cluster")

db = client["northstar_db"]
print("MongoDB connected:", client.server_info()['version'])
print("Database:", db.name)

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.8/1.8 MB 32.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 331.1/331.1 kB 15.5 MB/s eta 0:00:00
MongoDB connected: 8.0.23
Database: northstar_db


In [5]:
# Build Customer Documents
# Merge all data together to build customer documents
documents = []

for _, customer in customers.iterrows():
    cust_orders = orders[orders['customer_id'] == customer['customer_id']]

    order_list = []
    for _, order in cust_orders.iterrows():
        cust_deliveries = deliveries[deliveries['order_id'] == order['order_id']]
        cust_complaints = complaints[complaints['order_id'] == order['order_id']]
        cust_events = app_events[app_events['order_id'] == order['order_id']]

        delivery_data = cust_deliveries.to_dict('records')
        complaint_data = cust_complaints.to_dict('records')
        event_data = cust_events.to_dict('records')

        order_list.append({
            "order_id": order['order_id'],
            "service_type": order['service_type'],
            "pickup_zone": order['pickup_zone'],
            "dropoff_zone": order['dropoff_zone'],
            "order_value": order['order_value'],
            "deliveries": delivery_data,
            "complaints": complaint_data,
            "app_events": event_data
        })

    documents.append({
        "_id": customer['customer_id'],
        "age": customer['age'],
        "home_zone": customer['home_zone'],
        "customer_type": customer['customer_type'],
        "loyalty_score": customer['loyalty_score'],
        "orders": order_list
    })

print(f"Built {len(documents)} customer documents")

Built 650 customer documents


In [6]:
# Insert Documents into MongoDB
# Insert documents into MongoDB
collection = db["customer_journeys"]

# Clear collection first in case of re-runs
collection.drop()

# Insert all documents
collection.insert_many(documents)

print(f"Inserted: {collection.count_documents({})} documents")

Inserted: 650 documents


In [21]:
# CRUD Operations
# CREATE - Insert a new customer
new_customer = {
    "_id": "C9999",
    "age": 30,
    "home_zone": "North",
    "customer_type": "Consumer",
    "loyalty_score": 75.0,
    "orders": []
}

collection.insert_one(new_customer)
print("New customer inserted successfully")

New customer inserted successfully


In [8]:
# READ Operation
# READ - Find all customers in North zone with failed deliveries
result = collection.find({
    "home_zone": "North",
    "orders.deliveries.delivery_status": "Failed"
})

count = 0
for doc in result:
    print(f"Customer: {doc['_id']}, Zone: {doc['home_zone']}, Type: {doc['customer_type']}")
    count += 1

print(f"\nTotal customers found: {count}")

Customer: C0093, Zone: North, Type: Consumer
Customer: C0109, Zone: North, Type: Consumer
Customer: C0368, Zone: North, Type: Consumer
Customer: C0438, Zone: North, Type: Consumer
Customer: C0471, Zone: North, Type: Consumer
Customer: C0577, Zone: North, Type: Consumer
Customer: C0613, Zone: North, Type: Consumer
Customer: C0628, Zone: North, Type: Consumer

Total customers found: 8


In [9]:
# UPDATE Operation
# UPDATE - Update complaint status to Resolved
result = collection.update_many(
    {"orders.complaints.status": "Open"},
    {"$set": {"orders.$[].complaints.$[].status": "Resolved"}}
)

print(f"Documents matched: {result.matched_count}")
print(f"Documents updated: {result.modified_count}")

Documents matched: 52
Documents updated: 52


In [10]:
# DELETE - Remove the test customer we added earlier
result = collection.delete_one({"_id": "C9999"})

print(f"Documents deleted: {result.deleted_count}")
print(f"Total documents remaining: {collection.count_documents({})}")

Documents deleted: 1
Total documents remaining: 650


In [11]:
# Aggregation Query 1: Failed Deliveries by Zone
# Aggregation - Failed deliveries by zone
pipeline = [
    {"$unwind": "$orders"},
    {"$unwind": "$orders.deliveries"},
    {"$match": {"orders.deliveries.delivery_status": "Failed"}},
    {"$group": {
        "_id": "$home_zone",
        "total_failures": {"$sum": 1}
    }},
    {"$sort": {"total_failures": -1}}
]

results = collection.aggregate(pipeline)
print("Failed Deliveries by Zone:")
for r in results:
    print(f"Zone: {r['_id']}, Failures: {r['total_failures']}")

Failed Deliveries by Zone:
Zone: CENTRAL, Failures: 15
Zone: West, Failures: 12
Zone: EAST, Failures: 11
Zone: RiverSide, Failures: 10
Zone: SOUTH, Failures: 9
Zone: Riverside, Failures: 9
Zone: AIRPORT, Failures: 9
Zone: NORTH, Failures: 8
Zone: East, Failures: 8
Zone: North, Failures: 8
Zone: Airport, Failures: 8
Zone: north, Failures: 6
Zone: WEST, Failures: 6
Zone: South, Failures: 5
Zone: Central, Failures: 4
Zone: Ctr, Failures: 4


In [12]:
# Fix zone inconsistencies in MongoDB
collection.update_many(
    {},
    [{"$set": {"home_zone": {"$toUpper": "$home_zone"}}}]
)

print("Zone names standardised in MongoDB")

# Run aggregation again
pipeline = [
    {"$unwind": "$orders"},
    {"$unwind": "$orders.deliveries"},
    {"$match": {"orders.deliveries.delivery_status": "Failed"}},
    {"$group": {
        "_id": "$home_zone",
        "total_failures": {"$sum": 1}
    }},
    {"$sort": {"total_failures": -1}}
]

results = collection.aggregate(pipeline)
print("\nFailed Deliveries by Zone:")
for r in results:
    print(f"Zone: {r['_id']}, Failures: {r['total_failures']}")

Zone names standardised in MongoDB

Failed Deliveries by Zone:
Zone: NORTH, Failures: 22
Zone: RIVERSIDE, Failures: 19
Zone: EAST, Failures: 19
Zone: CENTRAL, Failures: 19
Zone: WEST, Failures: 18
Zone: AIRPORT, Failures: 17
Zone: SOUTH, Failures: 14
Zone: CTR, Failures: 4


After standardising zone names, North zone shows the highest
number of failed deliveries at 22, consistent with findings
from SQL and Python analysis. This cross-system consistency
strengthens the evidence that North zone requires immediate
operational attention.

In [13]:
# Aggregation Query 2: High Severity Open Complaints
# Aggregation - Customers with high severity open complaints
pipeline = [
    {"$unwind": "$orders"},
    {"$unwind": "$orders.complaints"},
    {"$match": {
        "orders.complaints.severity": "High",
        "orders.complaints.status": "Open"
    }},
    {"$group": {
        "_id": "$home_zone",
        "total_high_complaints": {"$sum": 1}
    }},
    {"$sort": {"total_high_complaints": -1}}
]

results = collection.aggregate(pipeline)
print("High Severity Open Complaints by Zone:")
for r in results:
    print(f"Zone: {r['_id']}, High Complaints: {r['total_high_complaints']}")

High Severity Open Complaints by Zone:


In [14]:
# Aggregation - Customers with high severity complaints
pipeline = [
    {"$unwind": "$orders"},
    {"$unwind": "$orders.complaints"},
    {"$match": {
        "orders.complaints.severity": "High"
    }},
    {"$group": {
        "_id": "$home_zone",
        "total_high_complaints": {"$sum": 1}
    }},
    {"$sort": {"total_high_complaints": -1}}
]

results = collection.aggregate(pipeline)
print("High Severity Complaints by Zone:")
for r in results:
    print(f"Zone: {r['_id']}, High Complaints: {r['total_high_complaints']}")

High Severity Complaints by Zone:
Zone: NORTH, High Complaints: 13
Zone: EAST, High Complaints: 12
Zone: RIVERSIDE, High Complaints: 11
Zone: WEST, High Complaints: 10
Zone: SOUTH, High Complaints: 10
Zone: CENTRAL, High Complaints: 10
Zone: AIRPORT, High Complaints: 9
Zone: CTR, High Complaints: 2


North zone leads in high severity complaints at 13, followed
by East at 12 and Riverside at 11. This is consistent with
findings from Python and SQL analysis, strongly suggesting
North zone has deep rooted operational problems affecting
customer satisfaction.

In [15]:
#Query Optimisation
# Create indexes
collection.create_index("home_zone")
collection.create_index("orders.complaints.severity")
collection.create_index([("home_zone", 1), ("orders.complaints.severity", 1)])

print("Indexes created successfully")
print("All indexes:", collection.index_information().keys())

Indexes created successfully
All indexes: dict_keys(['_id_', 'home_zone_1', 'orders.complaints.severity_1', 'home_zone_1_orders.complaints.severity_1'])


In [19]:
# Test query performance with explain()
result = collection.find(
    {"home_zone": "NORTH"}
).explain()

print("Query Performance Stats:")
print(f"Documents Examined: {result['executionStats']['totalDocsExamined']}")
print(f"Execution Time (ms): {result['executionStats']['executionTimeMillis']}")
print(f"Total Keys Examined: {result['executionStats']['totalKeysExamined']}")
print(f"Winning Plan Stage: {result['queryPlanner']['winningPlan']['stage']}")

Query Performance Stats:
Documents Examined: 111
Execution Time (ms): 0
Total Keys Examined: 111
Winning Plan Stage: FETCH


The explain() output shows the query examined 111 documents
using the home_zone index. The winning plan stage is FETCH
which confirms the index is being used. This is significantly
more efficient than a full collection scan of 650 documents.

In [20]:
# Drop index temporarily to compare performance
collection.drop_index("home_zone_1")

# Run same query without index
result_no_index = collection.find(
    {"home_zone": "NORTH"}
).explain()

print("WITHOUT Index:")
print(f"Documents Examined: {result_no_index['executionStats']['totalDocsExamined']}")
print(f"Execution Time (ms): {result_no_index['executionStats']['executionTimeMillis']}")
print(f"Winning Plan Stage: {result_no_index['queryPlanner']['winningPlan']['stage']}")

# Recreate index
collection.create_index("home_zone")
print("\nIndex recreated successfully")

WITHOUT Index:
Documents Examined: 111
Execution Time (ms): 0
Winning Plan Stage: FETCH

Index recreated successfully


With a small dataset of 650 documents the performance difference
between indexed and non-indexed queries is minimal. However in
a production environment with millions of records indexes on
home_zone and complaint severity would significantly reduce
documents examined and improve query response times.